File: C++/countGoodNodes.cpp
#include<iostream>
using namespace std;
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

// Problem 1448: Count Good Nodes in Binary Tree
// Given the root, a node in the tree is Good if the path from the root to X are no nodes with a value greater than X.
// Return the number of good nodes
class Solution {
public:
    int goodNodes(TreeNode* root) {
        //Recursion.
        return findGood(root, root->val);
    }
    //Returns int
    //root = current node
    //max = current maximum value in path
    int findGood(TreeNode* root, int max){
        //first, determine if node is valid
        if (!root){
            return 0;
        }
        //Now, determine if the value is greater than or equal to the max
        if (root->val >= max){
            //If so, we add 1 then recursively go through the rest of the nodes
            return 1 + findGood(root->left,root->val) + findGood(root->right, root->val);
        }
        //If not, then we just continue through the rest of the tree
        else{
            return findGood(root->left, max)+findGood(root->right, max);
        }
    }
};

File: Python/countGoodNodes.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# Problem 1448: Count Good Nodes in Binary Tree
# Given the root, a node in the tree is Good if the path from the root to X are no nodes with a value greater than X.
# Return the number of good nodes
class Solution(object):
    def goodNodes(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        #Idea: This is just an recursion question
        
        count = self.findGood(root,root.val)
        return count
    
    #function: return the number of good nodes.
    #Root = node in binary tree
    #maximum = maximum value in current path
    #Returns int
    def findGood(self, root, maximum):
        #Default, return 0 if there is no node
        if not root:
            return 0
        #If the value is greater than or equal to the maximum, then it's a good node
        if root.val >=maximum:
            #We'll add a 1 then recursively go through the rest of the tree
            return 1+self.findGood(root.left,root.val)+ self.findGood(root.right,root.val)
        #If not good, then just continue through the tree
        else:
            return self.findGood(root.left,maximum)+self.findGood(root.right,maximum)